openhcl/page_pool_alloc/src/lib.rs

```diff
@@ -23,6 +23,261 @@ use std::sync::Arc;
+                    pfn_bias,
+                    size_pages,
+                    device_id,
+                    tag,
+                } => InnerState::Allocated {
+                    base_pfn,
+                    pfn_bias,
+                    size_pages,
+                    device_id,
+                    tag,
+                },
+                State::AllocatedPendingRestore { .. } => {
+                    panic!("should not save AllocatedPendingRestore")
+                }
+                State::Leaked {
+                    base_pfn,
+                    pfn_bias,
+                    size_pages,
+                    device_id,
+                    tag,
+                } => InnerState::Leaked {
+                    base_pfn,
+                    pfn_bias,
+                    size_pages,
+                    device_id,
+                    tag,
+                },
+            }
+        }
+    }
+
+    impl From<InnerState> for State {
+        fn from(state: InnerState) -> Self {
+            match state {
+                InnerState::Free {
+                    base_pfn,
+                    pfn_bias,
+                    size_pages,
+                } => State::Free {
+                    base_pfn,
+                    pfn_bias,
+                    size_pages,
+                },
+                InnerState::Allocated {
+                    base_pfn,
+                    pfn_bias,
+                    size_pages,
+                    device_id,
+                    tag,
+                } => State::AllocatedPendingRestore {
+                    base_pfn,
+                    pfn_bias,
+                    size_pages,
+                    device_id,
+                    tag,
+                },
+                InnerState::Leaked {
+                    base_pfn,
+                    pfn_bias,
+                    size_pages,
+                    device_id,
+                    tag,
+                } => State::Leaked {
+                    base_pfn,
+                    pfn_bias,
+                    size_pages,
+                    device_id,
+                    tag,
+                },
+            }
+        }
+    }
+
+    // TODO: It seems unfortunate to define this type which is the same as
+    // MemoryRangeWithNode in vm_topology.
+    #[derive(Protobuf)]
+    #[mesh(package = "openhcl.pagepool")]
+    struct MemoryRangeWithNode {
+        #[mesh(1)]
+        range: MemoryRange,
+        #[mesh(2)]
+        vnode: u32,
+    }
+
+    #[derive(Protobuf)]
+    #[mesh(package = "openhcl.pagepool")]
+    enum DeviceIdState {
+        #[mesh(1)]
+        Used(String),
+        #[mesh(2)]
+        Unassigned(String),
+        #[mesh(3)]
+        Leaked(String),
+    }
+
+    impl From<DeviceId> for DeviceIdState {
+        fn from(state: DeviceId) -> Self {
+            match state {
+                DeviceId::PendingRestore(name) => {
+                    panic!("should not save PendingRestore, device name: {name}")
```

Comment:
Does this imply a situation where we attempt to save again while restore is still in progress? Or this is just generic precaution?

Reply:
This implies you have a bug where you didn't finish restoring the pool before we attempted to save again. It's a precautionary check